### ISSUE TITLE
import/no-unresolved is not recognizing installed modules

### ISSUE THREAD

**Issue by TheJaredWilcurt** [import-js/eslint-plugin-import#2703]
Errors:
```
  1:39  error  Unable to resolve path to module '@vue/test-utils'  import/no-unresolved
  4:29  error  Unable to resolve path to module 'vuex'             import/no-unresolved
```
Code:
```js
import { createLocalVue, mount } from '@vue/test-utils';
import axios from 'axios';
import flushPromises from 'flush-promises';
import Vuex, { Store } from 'vuex';

import { mutations } from '@/store.js';

import App from '@/App.vue';

import { currentUser } from '@@/mockData/CommonMockData.js';
import mockUsersResponse from '@@/mockResponses/Users.js';
```

`vuex` is in the `dependencies` and `@vue/test-utils` is in the `devDependencies` in `package.json`. Both are installed and in the `node_modules` folder.

What's interesting is that they fail across all test files, but no other `node_module` imports fail. It's just these two.

`eslintrc.js`
```js
const path = require('path');

module.exports = {
  root: true,
  parserOptions: {
    parser: '@babel/eslint-parser',
    ecmaVersion: 2022,
    sourceType: 'module',
    requireConfigFile: false
  },
  extends: [
    'tjw-import'
  ],
  settings: {
    'import/resolver': {
      webpack: {
        config: {
          resolve: {
            alias: {
              '@': path.resolve(__dirname, 'src'),
              '@@': path.resolve(__dirname, 'tests')
            }
          }
        }
      }
    }
  }
};
```

`tjw-import` can be seen here:

* https://github.com/tjw-lint/eslint-config-tjw-import/blob/main/index.js

**Comment by masi** [comment]
Do the package.json files have an "export" key to map exports to file names?

**Comment by TheJaredWilcurt** [comment]
```js
// /node_modules/@vue/test-utils/package.json
...
  "exports": {
    ".": {
      "default": "./dist/vue-test-utils.js",
      "require": "./dist/vue-test-utils.js",
      "import": "./dist/vue-test-utils.esm.js"
    }
  },
...
```
```js
// /node_modules/vuex/package.json
...
  "exports": {
    ".": {
      "module": "./dist/vuex.esm.js",
      "require": "./dist/vuex.common.js",
      "import": "./dist/vuex.mjs"
    },
    "./": "./"
  },
...
```
The `flush-promises` and `axios` imports do not have a linting error, and do not have `exports` defined in their `package.json`.

**Comment by PierreJeanjacquot** [comment]
hey, I'm running into the same issue while importing `@multiformats/multiaddr`

the issue may comes from nested export of `"."`

`@vue/test-utils`
![image](https://user-images.githubusercontent.com/26487010/217327631-ae062410-ced6-4b15-9f43-71edc698212a.png)

`@multiformats/multiaddr`
![image](https://user-images.githubusercontent.com/26487010/217328261-38bf5e9d-307c-4a62-9a0b-5f769d0ad8bc.png)

this syntax look good according to nodejs doc https://nodejs.org/docs/latest-v16.x/api/packages.html#package-entry-points but is no resolved by `eslint-plugin-import`

**Comment by TheJaredWilcurt** [comment]
Updated issue title based on this new information.

**Comment by ljharb** [comment]
Do they have a "main", though?

This plugin uses `resolve`, which does not yet support "exports".

**Comment by TheJaredWilcurt** [comment]
Updated previous example to answer question

**Comment by PierreJeanjacquot** [comment]
Thank you for the head up @ljharb 
No `"main"` neither in `@multiformats/multiaddr`
I looked at the `resolve` source it seems neither the legacy  `"module"` nor `"exports"` is supported
From Node.js 12 having only `"exports"` is fine, `"main"` is only used as a fallback when the former is missing. Since the oldest Node.js maintained LTS is 14 `"exports"` should be the default resolution.

**Comment by FloV22** [comment]
issue occurs also with [avajs](https://github.com/avajs/ava)

**Comment by TorinFrancis** [comment]
This issue also occurs with [react-hook-form](https://github.com/react-hook-form/react-hook-form), which does have a `"main"` field. For some reason, the issue does not occur with webpack 4, only 5. The error in the debug log comes from an webpack dependency (that conflicts with a eslint-import-resolver-webpack dependency).
```
eslint-plugin-import:resolver:webpack Error during module resolution: Error: Package path . is not exported from package /Users/torinfrancis/projects/churro/node_modules/react-hook-form (see exports field in /Users/torinfrancis/projects/churro/node_modules/react-hook-form/package.json)
    at /Users/torinfrancis/projects/churro/node_modules/webpack/node_modules/enhanced-resolve/lib/ExportsFieldPlugin.js:104:7
```

**Comment by MattIPv4** [comment]
👋 Just noting that I ran into this with `graphql-request` as well, which only has `exports` and not `main`:

```
...
"exports": {
    ".": {
      "require": {
        "types": "./build/cjs/index.d.ts",
        "default": "./build/cjs/index.js"
      },
      "import": {
        "types": "./build/esm/index.d.ts",
        "default": "./build/esm/index.js"
      }
    }
  },
...
```

**Comment by ivorpeles** [comment]
Also happening with solid-js

**Comment by 13OnTheCode** [comment]
Until 2023, the latest version of Node.js, 20.4.0, has arrived, yet people still refuse to embrace ESM. Why is that?

**Comment by ljharb** [comment]
@13OnTheCode because it's not actually better.

**Comment by Raynos** [comment]
Yeah it's the missing `"main"` field which causes an error on `sql-template-tag` package which appears to be ESM only and only has an `exports` set in the package.json

**Comment by chawes13** [comment]
I'm getting this error when trying to upgrade to the latest`date-fns` package, which does have `main` defined in its package.json (in addition to `exports`). https://github.com/date-fns/date-fns/blob/a39b8058f777517e8040193ba597dff18765951a/package.json#L17

**Comment by ljharb** [comment]
@chawes13 indeed, that points to https://unpkg.com/browse/date-fns@3.1.0/index.js which should work fine. Can you file a new issue for this one, since it's not related to "exports" support?

**Comment by JounQin** [comment]
@ljharb See https://github.com/import-js/eslint-plugin-import/issues/1479

**Comment by sla89** [comment]
I have the same issue with [@nats-io/nats-core](https://www.npmjs.com/package/@nats-io/nats-core):
`Unable to resolve path to module '@nats-io/nats-core'  import/no-unresolved`

From their package.json:
```json
  "name": "@nats-io/nats-core",
  "version": "3.0.0-30",
  "files": [
    "lib/",
    "LICENSE",
    "README.md"
  ],
  "types": "./lib/mod.d.js",
  "exports": {
    ".": "./lib/mod.js",
    "./internal": "./lib/internal_mod.js"
  },
```

![Image](https://github.com/user-attachments/assets/25b7d577-8bf4-4d1f-a997-e08ea87a470e)

**Comment by TheJaredWilcurt** [comment]
@ljharb

Would switching to using this library for resolving imports/exports be possible?

* https://www.npmjs.com/package/resolve.exports

> A tiny (952b), correct, general-purpose, and configurable "exports" and "imports" resolver without file-system reliance

It could even be used as a fallback from what you are currently doing, so it is only called if the current checks failed before reporting an error.

**Comment by ljharb** [comment]
@TheJaredWilcurt unfortunately no, because it's not actually correct - there's at least 6 different permutations of the `exports` algorithm depending on the node version, and we need to support all of them or none of them.

**Comment by TheJaredWilcurt** [comment]
Is there any downside to supporting all of them? That seems like the clear choice, since we can't control which approach libraries choose.

**Comment by ljharb** [comment]
They're contradictory. A library would have to allow the node version, and thus the algorithm, to be configurable per invocation, which is what `resolve` will eventually do.

**Comment by hnryjms** [comment]
I've been able to get this working using [`resolver#pathFilter()`](https://github.com/browserify/resolve/tree/main?tab=readme-ov-file#resolvesyncid-opts) like this:

```js
settings: {
  "import/resolver": {
    node: {
      pathFilter(pkg, path, relativePath) {
        return pkg.exports?.[`./${relativePath}`]?.default
          ?? pkg.exports?.["."]?.default
          ?? pkg.main
          ?? relativePath;
      },
    },
  },
},
```
